cpg-infrastructure/developer.rst

```diff
@@ -0,0 +1,142 @@
+
+We bring all group memberships under cpg-infrastructure, this allows us to:
+
+- unwrap groups completely, hence we know the full list of users / accounts that need to access specific resources.
+- better version control and store history of group memberships. In the CPG's configuration, we have CODEOWNER policies to manage who can add / remove users from groups.
+
+
+In code, we have a wrapper around a group, which is a list of GroupMember's, which is either a user or itself another group. We track which cloud we need to create the group for. At anytime you can request a full list of users, however the ID component might be a pulumi output value, eg: you can't know the service-account ID before you create it.
+
+All dataset infrastructure methods can create groups, and add members to them, and at the very end, these group members are finalised, synced to the cloud resource, and any access group caches are updated.
+
+
+How does it actually get deployed?
+----------------------------------
+
+
+In cpg-infrastructure-private, we have a deploy.yaml GitHub action. This:
+
+- Installs the latest cpg-infrastructure package
+- Runs the pulumi up command:
+    - This internally runs the ``__main__.py`` file, which is the entry point for the pulumi program. This packages up all the configuration in the private repo into:
+
+        1. A ``CPGInfrastructureConfig`` object, which contains users.
+        2. A ``list[CPGDatasetConfig]`` objects, which contains the datasets to spin up.
+
+Puluim does the rest!
+
+
+Previewing the workflow locally
+-------------------------------
+
+Prerequisites:
+
+* Have installed:
+    * Pulumi
+    * google-cloud-sdk
+    * azure-cli
+* Get and export the pulumi access token from a CPG team-member
+* Have both `cpg-infrastructure` and `cpg-infrastructure-private` cloned locally
+* Set-up a python virtual environment, installed cpg-infrastructure.
+    * Usually better to `pip install -e .` in the cpg-infrastructure directory.
+
+Noting we set these environment variables:
+
+- ``PULUMI_EXPERIMENTAL``: So we can use the skip checkpoints feature
+- ``PULUMI_SKIP_CHECKPOINTS``: Allows us to skip checkpoints which take forever!
+
+Some extra notes:
+
+- The `--non-interactive` flag is important because the interactive mode breaks for how many resources we have
+- The `--diff` flag is important because it shows us what pulumi is going to change, and it's a bug: https://github.com/pulumi/pulumi/issues/12162
+
+.. code-block:: bash
+
+    pulumi login gs://cpg-pulumi-state/
+    pulumi stack init production
+
+    PULUMI_EXPERIMENTAL=true PULUMI_SKIP_CHECKPOINTS=true pulumi preview \
+        --non-interactive --diff -p 20
```

Comment:
Could be worth adding a section with a link to the docs for developing locally here? Though they aren't public right now. 
https://github.com/populationgenomics/cpg-infrastructure-private/blob/main/docs/development.md